samples/shrubbery/reader.cc

```diff
@@ -0,0 +1,128 @@
+      "check parsing",
+      wf_check_parser,
+      dir::topdown,
+      {
+        // An empty block followed by alternatives is ignored
+        ((T(Block) << End) * (T(Alt)[Alt])) >>
+          [](Match& _) { return _(Alt); },
+
+        (T(Block) << (T(Group)[Group] << T(Alt))) >>
+          [](Match& _) { return Seq << *_[Group]; },
+
+        // An empty group caused by a semicolon is ignored
+        (In(Semi) * ((T(Group) << End))) >>
+          [](Match&) { return Seq ^ ""; },
+
+        // Commas must separate (non-empty) groups
+        (T(Comma) << End)[Comma] >>
+          [](Match& _) { return err(_[Comma], "Comma does not separate groups"); },
+
+        (In(Comma) * (T(Group) << End)[Group]) >>
+          [](Match& _) { return err(_[Group], "Comma does not separate groups"); },
+
+        // A comma can only appear inside a paren, brace or bracket
+        ((--In(Paren, Brace, Bracket)) * T(Comma)[Comma]) >>
+          [](Match& _) { return err(_[Comma], "Commas can only separate groups in parentheses/braces/brackets"); },
+
+        // Opener-closer pairs must have comma-separated groups
+        (In(Paren, Brace, Bracket) * Any * Any)[Group] >>
+          [](Match& _) { return err(_[Group], "Groups in parentheses/braces/brackets must be comma separated"); },
+
+        // Opener-closer pairs cannot have semicolon-separated groups
+        (In(Paren, Brace, Bracket, Comma) * T(Semi))[Semi] >>
+          [](Match& _) { return err(_[Semi], "Semicolons cannot separate groups in parentheses/brackets/braces. Use commas."); },
+
+        // Blocks cannot be empty, except immediately under opener-closer pairs
+        // and as the only term in a top-level group
+        (--(In(Paren, Brace, Bracket, Comma, File))) * ((T(Group) << ((!T(Block))++ * (T(Block)[Block] << End)))) >>
+          [](Match& _) { return err(_[Block], "Blocks may not be empty"); },
+
+        In(File) * (T(Group) << (((!T(Block)) * (!T(Block))++ * (T(Block)[Block] << End) * End))) >>
+          [](Match& _) { return err(_[Block], "Blocks may not be empty"); },
+
+        // Alternatives cannot be empty
+        T(Alt)[Alt] << End >>
+          [](Match& _) { return err(_[Alt], "Alternatives may not be empty"); },
+      }
+    };
+  }
+
+  // Merge alternatives into a single Alt node containing blocks
+  PassDef merge_alternatives()
+  {
+    return {
+      "merge alternatives",
+      wf_alternatives,
+      dir::topdown,
+      {
+        // Move a trailing alternative into the preceding group but do not cross
+        // a comma or semi-colon
+        (--In(Comma, Semi) * T(Group)[Group] * (T(Group) << (T(Alt)[Alt]))) >>
+          [](Match& _) { return _(Group) << _(Alt); },
+
+        // Alternatives keep their contents in a block
+        (T(Alt)[Alt] << !T(Block)) >>
+          [](Match& _) { return Alt << (Block << *_[Alt]); },
+
+        // Merge sequence of alternatives into a single alternative
+        (T(Alt)[Alt] * (T(Alt) << T(Block)[Block])) >>
```

Comment:
Potential alternative to allow `dir::once`?

```c++
(T(Alt)[Alt] * (T(Alt) << T(Block)))++[Group] >>
  [](Match& _) {
    Node alt = _(Alt);
    for(auto& node : _[Group]){
      alt << node->front();
    }
    return alt;
  }
```

Reply:
(NB this requires my PR which changes `NodeRange` to a span)

Reply:
Having foreach loops for `NodeRange` looks very nice! Are the parentheses in the pattern off by one? Reading the code, I would have guessed that the `Group` capture should not include the `Alt` capture, but only the subsequent nodes (so that we can add their blocks under the `alt` node in the effect.